src/components/fields/Formula/util.tsx

```diff
@@ -0,0 +1,142 @@
+import LongTextDisplayCell from "@src/components/fields/LongText/DisplayCell";
+import RichTextDisplayCell from "@src/components/fields/RichText/DisplayCell";
+import UrlDisplayCell from "@src/components/fields/Url/DisplayCell";
+import NumberDisplayCell from "@src/components/fields/Number/DisplayCell";
+import CheckboxDisplayCell from "@src/components/fields/Checkbox/DisplayCell";
+import PercentageDisplayCell from "@src/components/fields/Percentage/DisplayCell";
+import RatingDisplayCell from "@src/components/fields/Rating/DisplayCell";
+import SliderDisplayCell from "@src/components/fields/Slider/DisplayCell";
+import ColorDisplayCell from "@src/components/fields/Color/DisplayCell";
+import GeoPointDisplayCell from "@src/components/fields/GeoPoint/DisplayCell";
+import DateDisplayCell from "@src/components/fields/Date/DisplayCell";
+import DateTimeDisplayCell from "@src/components/fields/DateTime/DisplayCell";
+import ImageDisplayCell from "@src/components/fields/Image/DisplayCell";
+import FileDisplayCell from "@src/components/fields/File/DisplayCell";
+import JsonDisplayCell from "@src/components/fields/Json/DisplayCell";
+import CodeDisplayCell from "@src/components/fields/Code/DisplayCell";
+import MarkdownDisplayCell from "@src/components/fields/Markdown/DisplayCell";
+import CreatedByDisplayCell from "@src/components/fields/CreatedBy/DisplayCell";
+
+export function useDeepCompareMemoize<T>(value: T) {
+  const ref = useRef<T>(value);
+  const signalRef = useRef<number>(0);
+
+  if (!isEqual(value, ref.current)) {
+    ref.current = value;
+    signalRef.current += 1;
+  }
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  return useMemo(() => ref.current, [signalRef.current]);
+}
+
+export const listenerFieldTypes = Object.values(FieldType).filter(
+  (type) =>
+    ![
+      FieldType.action,
+      FieldType.status,
+      FieldType.formula,
+      FieldType.aggregate,
+      FieldType.connectService,
+      FieldType.connectTable,
+      FieldType.connector,
+      FieldType.subTable,
+      FieldType.last,
+    ].includes(type)
+);
+
+export const outputFieldTypes = Object.values(FieldType).filter(
+  (type) =>
+    ![
+      FieldType.formula,
+      FieldType.derivative,
+      FieldType.action,
+      FieldType.status,
+      FieldType.aggregate,
+      FieldType.connectService,
+      FieldType.connectTable,
+      FieldType.connector,
+      FieldType.duration,
+      FieldType.subTable,
+      FieldType.reference,
+      FieldType.createdAt,
+      FieldType.createdBy,
+      FieldType.updatedAt,
+      FieldType.updatedBy,
+      FieldType.last,
+    ].includes(type)
+);
+
+export const defaultFn = `const formula:Formula = async ({ row })=> {
+  // Write your formula code here
+  // for example:
+  // return row.a + row.b;
+  // checkout the documentation for more info: https://docs.rowy.io/field-types/formula
+}
+`;
+
+export const getDisplayCell = (type: FieldType) => {
```

Comment:
I tried that first but that way there's no easy workaround to disable firestore sync because UI is highly coupled with logic.  

For the preview table feature, I will consider this as well regarding separate database logic from UI.